[RDY] Fix for missing output (#4569, ...)

oni-link opened the pull request:
process.c: Process termination closes streams too early 
 shell.c: Fix missing output 

Some tests fail for me, but so do the tests for upstream. Not sure what is wrong. Testing with Travis CI.

Fixes #4569

oni-link committed:
process.c: Prevent data loss for process output streams

For a terminating process, it's output streams could be closed,
before all data is read.

oni-link committed:
shell.c: Fix missing output

The whole stream buffer is now put on screen at once instead of only
data up to the last newline. This has some advantages:

* RBuffer cannot wrap around, so we never forget to output second
  half of the buffer.
* Stream data is not delayed anymore, because we don't have to wait for
  a newline.

This works by remembering the last used screen column.

oni-link commented:
Only test that fails/hangs is `test/functional/legacy/command_count_spec.lua`, but it also hangs on upstream.

justinmk commented:
Still WIP?

oni-link commented:
@justinmk, I wanted to remove the teardown check [here](https://github.com/neovim/neovim/pull/4646/files#diff-89f96344a6610f20b684b4d71368eaabR373), but without it, the event loop could be called recursively. So with the check in place, perhaps there could be some data loss when nvim exits and does not check the open streams.

I also have a local commit that wraps around long lines for `:!cmd` output, but this could wait for another PR.

Otherwise it is ready for RFC (and feedback how it works on different systems).

oni-link commented:
`tempfile_spec.lua` fix removed and some commits squashed.

justinmk commented:
@oni-link Any chance you could write a test for this? This fix is valuable, and we really don't want to regress it.

justinmk commented on src/nvim/event/process.c:
```diff
@@ -333,9 +333,47 @@ static void process_close(Process *proc)
   Process *proc = argv[0];
+
+  // Did our process forked a child that keeps the output streams open?
+  if (!process_is_tearing_down) {
```
How about a TODO here mentioning "there could be some data loss when nvim exits and does not check the open streams."

oni-link replied:
After commit fixup2 there should only be data loss for a child process that keeps the streams open.  Otherwise something like this
`:!./a.out` with `a.out` from
```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char *argv[])
{
  pid_t cpid = fork();

  if (cpid == -1) {
    perror("fork");
    exit(EXIT_FAILURE);
  }

  if (cpid == 0) {
    char* arg[]={"/usr/bin/yes",NULL};
    execv("/usr/bin/yes",arg);

    _exit(EXIT_SUCCESS);

  } else { 
    sleep(3);
    exit(EXIT_SUCCESS);
  }
}

```
would block forever and nvim could not exit (without data loss).

oni-link committed:
fixup: process.c: Prevent data loss for process output streams

* Get system buffer size for upper data limit. Otherwise data loss
  if this buffer is too big.
* Test whether teardown needs special handling.

oni-link commented:
> @oni-link Any chance you could write a test for this? This fix is valuable, and we really don't want to regress it.

I tried, but I did not get very far (screen test problems). My manuel tests were

`:!cat file.txt` " in upstream second half of output is lost
`:so vimtest.vim` " in upstream second half of output is lost
`:call system('xclip -i -selection clipboard','data')` " check that changes can handle child process that keeps the streams open after parent terminated


I used `file.txt` from [#4569](https://github.com/neovim/neovim/files/217609/nvim_test.zip) and modified script  `vimtest.vim`
```vim
function! s:JobHandler(job_id, data, event)
  echo a:job_id .", ".string(a:data) .", " .a:event
  endfunction

let s:callbacks = {
    \ 'on_stdout': function('s:JobHandler'),
    \ 'on_stderr': function('s:JobHandler'),
    \ 'on_exit': function('s:JobHandler')
    \ }

let job1 = jobstart(['bash', '-c', 'cat file.txt'], s:callbacks)

```

oni-link committed:
fixup2: process.c: Prevent data loss for process output streams

The only data loss should be, if a process forked a child that keeps
sending data after the parent terminated.
While not in teardown mode we could keep reading child data, but then
`:!cmd` would block after `cmd` exited. In teardown mode we want to exit
nvim so we cannot keep reading child data.

justinmk commented:
@oni-link I also noticed that wrapping the last line in `jobwait()` avoids the issue:

    let job1 = jobwait([jobstart(['bash', '-c', 'cat file.txt'], s:callbacks)], 3000)

However, I finally found a way to repro the issue in a test, by spawning a child nvim in a `:terminal` buffer. Working on a PR.

justinmk commented on src/nvim/rbuffer.c:
```diff
@@ -15,7 +15,7 @@ RBuffer *rbuffer_new(size_t capacity)
 {
   if (!capacity) {
-    capacity = 0xffff;
+    capacity = 0x10000;
```
@oni-link Can you explain this change?

oni-link replied:
Suggested buffer size by libuv and linux pipe capacity (also a multiple of the atomic pipe buffer size (4096)).

justinmk commented:
Though https://github.com/neovim/neovim/pull/4950 didn't help on CI, it fails reliably on my machine (before your PR). 

Is this PR ready?

oni-link commented:
I cannot reproduce any fails when running `TEST_FILE="./test/functional/terminal/job_spec.lua"  make functionaltest` when using #4950 (without this PR).

justinmk commented:
> I cannot reproduce any fails when running TEST_FILE="./test/functional/terminal/job_spec.lua" make functionaltest when using #4950 (without this PR).

Neither can travis, or quickbuild. It only seems to fail on a single machine I've tried. I'll take another look at it later.
